src/H5Dmpio.c

```diff
@@ -5497,9 +5500,6 @@ H5D__mpio_collective_filtered_vec_io(const H5D_filtered_collective_io_info_t *ch
              */
             assert(vec_idx < iovec_count);
 
-            if (op_type == H5D_IO_OP_READ && !chunk_list->chunk_infos[i].need_read)
-                continue;
```

Comment:
The logic here needed to first check whether the chunk in question is actually going to be read before asserting that `vec_idx < iovec_count` to prevent buffer overruns. If the last chunk(s) in this array don't need to be read, we will have `vec_idx == iovec_count`, but that's ok because we'll skip adding the remaining chunks to the I/O vector buffers.